src/Core/Scheduler/Providers/TileProvider.js

```diff
@@ -9,6 +9,31 @@ import TileGeometry from '../../TileGeometry';
 import TileMesh from '../../TileMesh';
 import CancelledCommandException from '../CancelledCommandException';
 import { requestNewTile } from '../../../Process/TiledNodeProcessing';
+import RendererConstant from '../../../Renderer/RendererConstant';
+import Picking from '../../Picking';
+
+function changeRenderState(tileLayer) {
```

Comment:
All references to the layer should disappear from `tileProvider`. Because there isn't reason for a tile provider to know the layers. For example, in the future, remove any notion of layer in `TileProvider.prototype.executeCommand`. You have to be able to use the tile provider without a layer concept. 

`preprocessDataLayer` also has no reason to be here and even less if that add functions that aren't `tile building`

So don't add new reverence to the layer

Reply:
> Because there isn't reason for a tile provider to know the layers

Every single provider of itowns absolutely *does* know about layers. Providers are here to interface layers with 2 things: 
- a protocol by:
    - initializing the layer with some default values for this protocol and / or metadata from the network, following the protocol specifications (getCapabilities, tileset.json...)
    - abstract the fetching of data from servers implementing this protocol
- The scheduler mechanism, because we're fetching datas from the internet. And it uses the layer because it provides some configurations that `executeCommand` needs to generate new nodes (opacity, colors, overrideMaterials etc...)

And in turns, the way we create node for a given protocol have an influence on how we should pick it on the scene. So it makes a lot of sense that this logic should be at least partly in providers.

> preprocessDataLayer also has no reason to be here

For the aforementioned reason, it has all the reason to be in a provider

Not sure I understand you well. Could you describe what alternative solution you would choose? Where would you move all that code then? It might make things clearer.

Reply:
Moved more code so now TileProvider isn't modified.

Is this what you had in mind @gchoqueux ?

Reply:
The best place for picking is the construction of the geometric layer using the three.js raycaster caster picking. And then overload it in more specific cases like the tile layer. In this way, there is no need to add the picking in each provider of geometry.
@autra  It's clearer?

Reply:
Somehow... 

> And then overload it in more specific cases like the tile layer

Where would you overload it then? I found that providers was a good place for that (because they constructs the objects, so they know how they should be picked). That being said I'm fine with the current version.